src/compact_cow_str.rs

```diff
@@ -0,0 +1,243 @@
+pub struct CompactCowStr<'a> {
+    // `tagged_len` is a tag in its highest bit, and the string length in the rest of the bits.
+    //
+    // * If the tag is 1, the memory pointed to by `ptr` is owned
+    //   and the lifetime parameter is irrelevant.
+    //   `ptr` and `len` are the components of a `Box<str>`.
+    //
+    // * If the tag is 0, the memory is borrowed.
+    //   `ptr` and `len` are the components of a `&'a str`.
+
+    ptr: *const u8,
+    tagged_len: usize,
+    phantom: PhantomData<&'a str>,
+}
+
+impl<'a> From<&'a str> for CompactCowStr<'a> {
+    #[inline]
+    fn from(s: &'a str) -> Self {
+        let len = s.len();
+        assert!(len <= MAX_LEN);
+        CompactCowStr {
+            ptr: s.as_ptr(),
+            tagged_len: len,
+            phantom: PhantomData,
+        }
+    }
+}
+
+impl<'a> From<Box<str>> for CompactCowStr<'a> {
+    #[inline]
+    fn from(s: Box<str>) -> Self {
+        let ptr = s.as_ptr();
+        let len = s.len();
+        assert!(len <= MAX_LEN);
+        mem::forget(s);
+        CompactCowStr {
+            ptr: ptr,
+            tagged_len: len | TAG,
+            phantom: PhantomData,
+        }
+    }
+}
+
+impl<'a> CompactCowStr<'a> {
+    /// Wether this string refers to borrowed memory
+    /// (as opposed to owned, which would be freed when `CompactCowStr` goes out of scope).
+    #[inline]
+    pub fn is_borrowed(&self) -> bool {
+        (self.tagged_len & TAG) == 0
+    }
+
+    /// The length of this string
+    #[inline]
+    pub fn len(&self) -> usize {
+        self.tagged_len & !TAG
+    }
+
+    // Intentionally private since it is easy to use incorrectly.
+    #[inline]
+    fn as_raw_str(&self) -> *const str {
+        unsafe {
+            str::from_utf8_unchecked(slice::from_raw_parts(self.ptr, self.len()))
+        }
+    }
+
+    /// If this string is borrowed, return a slice with the original lifetime,
+    /// not borrowing `self`.
+    ///
+    /// (`Deref` is implemented unconditionally, but returns a slice with a shorter lifetime.)
+    #[inline]
+    pub fn as_str(&self) -> Option<&'a str> {
+        if self.is_borrowed() {
+            Some(unsafe { &*self.as_raw_str() })
+        } else {
+            None
+        }
+    }
+
+    /// Convert into `String`, re-using the memory allocation it was already owned.
+    #[inline]
+    pub fn into_owned(self) -> String {
+        unsafe {
+            let raw = self.as_raw_str();
+            let is_borrowed = self.is_borrowed();
+            mem::forget(self);
+            if is_borrowed {
+                String::from(&*raw)
+            } else {
+                Box::from_raw(raw as *mut str).into_string()
+            }
+        }
+    }
+}
+
+impl<'a> Clone for CompactCowStr<'a> {
+    #[inline]
+    fn clone(&self) -> Self {
+        if self.is_borrowed() {
+            CompactCowStr { ..*self }
+        } else {
+            Self::from(String::from(&**self).into_boxed_str())
```

Comment:
If `self` is borrowed, the new value will be borrowed too. If not, the lifetime parameter is irrelevant: it could be `'static`, and the owned memory does not necessarily live that long.